### PR TITLE
[GRPC][Part 8] Propagate Application Headers

### PR DESCRIPTION
**Pull request by willhug**
Summary: Adds the ability for GRPC to send application headers.  Adds a
header file for containing the process of converting headers between
grpc and yarpc

Test Plan: Wrote tests, added to crossdock header test, ran
yclient/yserver

### CODE REVIEW

File: transport/grpc/handler_test.go [hunk 1]
```diff
@@ -3,10 +3,13 @@ package grpc
 import (
 	"testing"
 
+	"google.golang.org/grpc/metadata"
+
```
Comment (abhinav) [inline transport/grpc/handler_test.go]: import groups

File: transport/grpc/header.go [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+	applicationHeaders = headerMapper{ApplicationHeaderPrefix}
+)
+
+// toGRPCMetadata converts transport headers into gRPC Metadata.
+//
+// Headers are read from 'from' and written to 'to'. The final header collection
+// is returned.
+//
+// If 'to' is nil, a new map will be assigned.
+func (hm headerMapper) ToGRPCMetadata(from transport.Headers, to metadata.MD) metadata.MD {
```
Comment (abhinav) [inline transport/grpc/header.go]: No need to name this `ToGRPCMetadata`. Just `ToMetadata` should be fine.

Also, don't make this public. This shouldn't be part of our public API. Move it to an internal package if necessary.

File: transport/grpc/header.go [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+	return to
+}
+
+// fromGRPCMetadata converts GRPC Metadata to transport headers.
+//
+// Headers are read from 'from' and written to 'to'. The final header collection
+// is returned.
+//
+// If 'to' is nil, a new map will be assigned.
+func (hm headerMapper) FromGRPCMetadata(from metadata.MD, to transport.Headers) transport.Headers {
```
Comment (abhinav) [inline transport/grpc/header.go]: Same as above

File: transport/grpc/header.go [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+			key := k[prefixLen:]
+			to = to.With(key, Headers(from).Get(k))
+		}
+		// Note: undefined behavior for multiple occurrences of the same header
+	}
+	return to
+}
+
+// Headers is a convenience type for converting Header information safely
+type Headers map[string][]string
```
Comment (abhinav) [inline transport/grpc/header.go]: Why define this? Do we really want people to use `grpc.Headers` explicitly?

Comment (willhug) [inline transport/grpc/header.go]: Hmm, yeah this didn't need to be external, it was added as a simplification step for working on metadata headers, but we can just define the functions on metadata headers themselves.

File: transport/grpc/handler_test.go [hunk 5]
```diff
@@ -64,3 +67,32 @@ func TestResponseWriter_Write(t *testing.T) {
 	assert.Equal(t, error(nil), err)
 	assert.Equal(t, strMsg, string(r.body.Bytes()))
 }
+
+func TestResponseWriter_AddHeaders(t *testing.T) {
```
Comment (abhinav) [inline transport/grpc/handler_test.go]: We don't generally add underscores to test names. Just use
`TestResponseWriterWithHeaders` or something similar.

File: transport/grpc/handler_test.go [hunk 6]
```diff
@@ -64,3 +67,32 @@ func TestResponseWriter_Write(t *testing.T) {
...
+
+	var r response
+	rw := newResponseWriter(&r)
+
+	rw.AddHeaders(inputHeaders)
+
+	assert.Equal(t, expectedHeaders, r.headers)
+}
+
+func TestResponseWriter_SetApplicationError(t *testing.T) {
```
Comment (abhinav) [inline transport/grpc/handler_test.go]: TestResponseWriterApplicationError

File: transport/grpc/handler_test.go [hunk 7]
```diff
@@ -64,3 +67,32 @@ func TestResponseWriter_Write(t *testing.T) {
...
+	var r response
+	rw := newResponseWriter(&r)
+
+	rw.SetApplicationError()
+
+	// No action on Application Error
```
Comment (abhinav) [inline transport/grpc/handler_test.go]: If there's no behavioral change, I wouldn't even bother with testing it. If
you just want to make sure that that code path is exercised, crossdock will
take care of it.

File: transport/grpc/header.go [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+
+	"google.golang.org/grpc/metadata"
+)
+
+// headerMapper converts gRPC Metadata to and from transport headers.
+type headerMapper struct{ Prefix string }
+
+var (
+	applicationHeaders = headerMapper{ApplicationHeaderPrefix}
+)
```
Comment (abhinav) [inline transport/grpc/header.go]: nit: No need for a group if there's only one entry

```
var applicationHeaders = headerMapper{..}
```

File: transport/grpc/header.go [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+		if strings.HasPrefix(k, hm.Prefix) {
+			key := k[prefixLen:]
+			to = to.With(key, headers(from).get(k))
+		}
+		// Note: undefined behavior for multiple occurrences of the same header
+	}
+	return to
+}
+
+// headers is a convenience type for converting Header information safely
```
Comment (abhinav) [inline transport/grpc/header.go]: You say 'safely' but you're not normalizing the keys. Don't you need to do that?

File: transport/grpc/header_test.go [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+		"key1": "value1",
+		"key2": "value2",
+	})
+
+	actualHeaders := testMapper.fromMetadata(inputMetadata, transport.Headers{})
+
+	assert.Equal(t, expectedHeaders, actualHeaders)
+}
+
+func TestHeaders_Add(t *testing.T) {
```
Comment (abhinav) [inline transport/grpc/header_test.go]: Remove `_` from all test names please

**Comment by breerly** [comment]
How are we matching up to the GRPC headers paradigm? Is this going to put us in a place where setting headers on a pure GRPC client don't make it to a YARPC server?

File: transport/grpc/header.go (hunk 9 above)
Comment (willhug) [inline transport/grpc/header.go]: I'll fix that

**Comment by willhug** [comment]
@breerly since we're using the GRPC metadata to pass our application headers we should be adhering to all of their paradigms since this is the public API that they surface, anything else that happens header-wise will be done by GRPC under the hood.  The only thing that might cause us some issues is name we've given to our "passthrough" encoding, which we'll need middleware to solve on the pure GRPC servers.

**Comment by abhinav** [comment]
According to http://www.grpc.io/docs/guides/wire.html, GRPC sends application
headers as-is. It just reserves the `grpc-*` namespace of headers for internal
use. So a GRPC server receiving a request from a YARPC client will have to
strip the `rpc-header-` prefix.

Should we maybe send application headers as the user specified them in the
GRPC transport? The GRPC inbound will have to pass-through all Metadata as
YARPC headers except the `RPC-*` headers.

**Pull request merged**